subprojects/dependency-locking/src/integTest/groovy/org/gradle/dependency/locking/DependencyLockingPluginIntegrationTest.groovy

```diff
@@ -0,0 +1,165 @@
+ */
+
+package org.gradle.dependency.locking
+
+import org.gradle.integtests.fixtures.AbstractDependencyResolutionTest
+
+import static org.gradle.util.Matchers.containsText
+
+class DependencyLockingPluginIntegrationTest extends AbstractDependencyResolutionTest {
+
+    def 'succeeds with lock file present'() {
+        mavenRepo.module('org', 'foo', '1.0').publish()
+
+        buildFile << """
+apply plugin: 'dependency-locking'
+
+repositories {
+    maven {
+        name 'repo'
+        url '${mavenRepo.uri}'
+    }
+}
+configurations {
+    lockedConf
+}
+
+dependencies {
+    lockedConf 'org:foo:1.+'
+}
+"""
+
+        file('gradle', 'dependency-locks', 'lockedConf.lockfile') << """
+org:foo:1.0
+"""
+        expect:
+        succeeds 'dependencies'
+    }
+
+    def 'fails with out-of-date lock file'() {
+        mavenRepo.module('org', 'foo', '1.0').publish()
+        mavenRepo.module('org', 'foo', '1.1').publish()
+
+        buildFile << """
+apply plugin: 'dependency-locking'
+
+repositories {
+    maven {
+        name 'repo'
+        url '${mavenRepo.uri}'
+    }
+}
+configurations {
+    lockedConf
+}
+
+dependencies {
+    constraints {
+        lockedConf('org:foo') {
+            version {
+                prefer '1.1'
+            }
+        }
+    }
+    lockedConf 'org:foo:1.+'
+}
+"""
+
+        file('gradle', 'dependency-locks', 'lockedConf.lockfile') << """
+org:foo:1.0
+"""
+        when:
+        fails 'dependencies'
+
+        then:
+//        failure.assertThatCause(containsText(LockOutOfDateException.class.name)) // Fails to find the exception name in the cause
+        failure.assertThatCause(containsText("Lock file expected 'org:foo:1.0' but resolution result was 'org:foo:1.1'"))
+    }
+
+    def 'fails when lock file entry not resolved'() {
+        mavenRepo.module('org', 'foo', '1.0').publish()
+        mavenRepo.module('org', 'bar', '1.0').publish()
+
+        buildFile << """
+apply plugin: 'dependency-locking'
+
+repositories {
+    maven {
+        name 'repo'
+        url '${mavenRepo.uri}'
+    }
+}
+configurations {
+    lockedConf
+}
+
+dependencies {
+    lockedConf 'org:foo:1.+'
+}
+"""
+
+        file('gradle', 'dependency-locks', 'lockedConf.lockfile') << """
+org:bar:1.0
+org:foo:1.0
+"""
+        when:
+        fails 'dependencies'
+
+        then:
+        failure.assertThatCause(containsText("Lock file contained module 'org:bar:1.0' but it is not part of the resolved modules"))
+    }
+
+    def 'writes dependency lock file when requested'() {
+        mavenRepo.module('org', 'foo', '1.0').publish()
+        mavenRepo.module('org', 'bar', '1.0').publish()
+
+        buildFile << """
+apply plugin: 'dependency-locking'
+
+repositories {
+    maven {
+        name 'repo'
+        url '${mavenRepo.uri}'
+    }
+}
+configurations {
+    lockedConf
+}
+
+dependencies {
+    lockedConf 'org:foo:1.+'
+    lockedConf 'org:bar:1.+'
+}
+
+tasks.dependencies.dependsOn saveDependencyLocks
```

Comment:
While this says "save the locks if I run the _dependencies_ task", it also says "first save the locks, then run the _dependencies_ task". But shouldn't it be the other way around? First run the task that resolves the configurations, then save the locks at the end?

Reply:
That is just a temporary hack because I did not handle task ordering at this stage. The goal is to make this totally irrelevant in the end.